src/main/groovy/com/morpheus/scvmm/sync/TemplatesSync.groovy

```diff
@@ -0,0 +1,613 @@
+package com.morpheus.scvmm.sync
+
+import com.morpheus.scvmm.ScvmmApiService
+import com.morpheusdata.core.MorpheusContext
+import com.morpheusdata.core.data.DataFilter
+import com.morpheusdata.core.data.DataOrFilter
+import com.morpheusdata.core.data.DataQuery
+import com.morpheusdata.core.providers.CloudProvider
+import com.morpheusdata.core.util.ComputeUtility
+import com.morpheusdata.core.util.SyncTask
+import com.morpheusdata.model.*
+import com.morpheusdata.model.projection.StorageVolumeIdentityProjection
+import com.morpheusdata.model.projection.VirtualImageIdentityProjection
+import com.morpheusdata.model.projection.VirtualImageLocationIdentityProjection
+import groovy.util.logging.Slf4j
+import io.reactivex.rxjava3.core.Observable
+
+/**
+ * @author rahul.ray
+ */
+
+@Slf4j
+class TemplatesSync {
+    private Cloud cloud
+    private ComputeServer node
+    private MorpheusContext context
+    private ScvmmApiService apiService
+    private CloudProvider cloudProvider
+
+    TemplatesSync(Cloud cloud, ComputeServer node, MorpheusContext context, CloudProvider cloudProvider) {
+        this.cloud = cloud
+        this.node = node
+        this.context = context
+        this.apiService = new ScvmmApiService(context)
+        this.@cloudProvider = cloudProvider
+
+    }
+
+    def execute() {
+        log.debug "TemplatesSync"
+        def scvmmOpts = apiService.getScvmmZoneAndHypervisorOpts(context, cloud, node)
+        def listResults = apiService.listTemplates(scvmmOpts)
+        if (listResults.success && listResults.templates) {
+            def existingLocations = context.services.virtualImage.location.listIdentityProjections(new DataQuery()
+                    .withFilter('refType', 'ComputeZone')
+                    .withFilter('refId', cloud.id)
+                    .withFilter('virtualImage.imageType', 'in', ['vhd', 'vhdx'])
+                    .withJoin('virtualImage'))
+            def groupedLocations = existingLocations.groupBy({ row -> row.externalId })
+            def dupedLocations = groupedLocations.findAll { key, value -> value.size() > 1 }
+            def dupeCleanup = []
+            if (dupedLocations?.size() > 0)
+                log.warn("removing duplicate image locations: {}", dupedLocations.collect { it.key })
+            dupedLocations?.each { key, value ->
+                value.eachWithIndex { row, index ->
+                    if (index > 0)
+                        dupeCleanup << row
+                }
+            }
+            dupeCleanup?.each { row ->
+                def dupeResults = context.async.virtualImage.location.remove([row.id]).blockingGet()
+                if (dupeResults == true) {
+                    existingLocations.remove(row)
+                }
+            }
+            def domainRecords = Observable.fromIterable(existingLocations)
```

Comment:
this will cause a slow down in sync. Do the deduplication first and then execute the identity projection query again. If you want to share the query filters just reuse the DataQuery and filters in 2 separate `listIdentityProjection` calls.